lib/Sema/TypeCheckDecl.cpp

```diff
@@ -7946,6 +7946,67 @@ static void diagnoseClassWithoutInitializers(TypeChecker &tc,
   tc.diagnose(classDecl, diag::class_without_init,
               classDecl->getDeclaredType());
 
+  // HACK: We've got a special case to look out for and diagnose specifically to
+  // improve the experience of seeing this, and mitigate some confusion.
+  //
+  // For a class A which inherits from Decodable class B, class A may have
+  // additional members which prevent default initializer synthesis (and
+  // inheritance of other initializers). The user may have assumed that this
+  // case would synthesize Encodable/Decodable conformance for class A the same
+  // way it may have for class B, or other classes.
+  //
+  // It is helpful to suggest here that the user may have forgotten to override
+  // init(from:) (and encode(to:), if applicable) in a note, before we start
+  // listing the members that prevented initializer synthesis.
+  // TODO: Add a fixit along with this suggestion.
+  if (auto *superclassDecl = classDecl->getSuperclassDecl()) {
+    ASTContext &C = tc.Context;
+    auto *decodableProto = C.getProtocol(KnownProtocolKind::Decodable);
+    auto superclassType = superclassDecl->getDeclaredInterfaceType();
+    if (auto ref = tc. conformsToProtocol(superclassType, decodableProto,
+                                          superclassDecl,
+                                          ConformanceCheckOptions(),
+                                          SourceLoc())) {
+      // super conforms to Decodable, so we've failed to inherit init(from:).
+      // Let's suggest overriding it here.
```

Comment:
Actually, we never hit this diagnostic in cases like this. If the class adds no stored properties (`SuppressDefaultInitializer` in `addImplicitConstructors` is false), we always try to inherit the superclass's initializers. If it has none (because synthesis failed or for any other reason), we simply don't inherit any, and never complain.

This is true in Swift 3 and Swift 4:

```swift
class Super {
	var value: Double
}

class Sub : Super {}
```

If we'd like to complain there, then I can add a check in both locations (whether we have no initializers and also inherited none from `super`, and also whether we're about to produce a diagnostic about something that's not correct); otherwise we're safe to leave things as-is.